bee-plugin/src/plugin.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Utilities to write plugins.
+
+pub use crate::handshake::HandshakeInfo;
+use crate::{
+    error::PluginError,
+    event::*,
+    grpc::{
+        plugin_server::{Plugin as GrpcPlugin, PluginServer},
+        ProcessReply, ShutdownReply, ShutdownRequest,
+    },
+};
+
+use tokio::io::{stdout, AsyncWriteExt};
+use tonic::{transport::Server, Request, Response, Status, Streaming};
+
+macro_rules! plugin_trait {
+    ($($method_name:ident => $event_ty:ty),*) => {
+        /// Types that represent a plugin.
+        #[tonic::async_trait]
+        pub trait Plugin: Send + Sync + 'static {
+            /// Returns the `HandshakeInfo` of the current plugin.
+            fn handshake_info() -> HandshakeInfo;
+            /// Prepares the plugin for shutdown.
+            async fn shutdown(&self);
+            $(
+                #[doc = concat!("Handles an event of type `", stringify!($event_ty), "`.")]
+                async fn $method_name(&self, _event: $event_ty) {}
+            )*
+        }
+
+        #[tonic::async_trait]
+        impl<T: Plugin> GrpcPlugin for T {
+            async fn shutdown(&self, _request: Request<ShutdownRequest>) -> Result<Response<ShutdownReply>, Status> {
+                self.shutdown().await;
+                Ok(Response::new(ShutdownReply {}))
+            }
+
+            $(
+                async fn $method_name(
+                    &self,
+                    request: Request<Streaming<$event_ty>>,
+                ) -> Result<Response<ProcessReply>, Status> {
+                    let mut stream = request.into_inner();
+
+                    while let Some(message) = stream.message().await? {
+                        self.$method_name(message).await;
+                    }
+
+                    Ok(Response::new(ProcessReply {}))
+                }
+            )*
+        }
+    };
+}
+
+plugin_trait! {
```

Comment:
Can we not use dynamic dispatch or even `Any` for the events (as we already do in the event bus)?

Reply:
maybe we can but I don't know if that's more newbie friendly because then you have to do a bunch of downcasting and so on to get a concrete event type (there's the performance argument too).